### PR TITLE
Bg test data #162437173

### PR DESCRIPTION
**Pull request by Muthembwa**
**What does this PR do?**

The PR Implements Post in OOP model, it also modifies the tests and adds edge cases 

I renamed V1 to v1 for ease of use, implemented a base_case file containing the test data models. 

**Related pivotal stories**
#162437173

vies
models
image

### CODE REVIEW

File: app/api/v1/views.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+
+class RedFlags(Resource):
+    def __init__(self,CreatedBy, Location, Status, Comment):
+        self.db = RedFlagModel()
+         
+
+    def post(self): 
+        daatedBy = data['CreatedBy']
```
Comment (giddygitau) [inline app/api/v1/views.py]: Where are we using the properties you extracted from the request? You are passing `data` directly to `RedFlagModel()` - which required 4 parameters instead of one

File: app/api/v1/views.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        Locta = request.get_json(force=True)
+        Creation = data['Location']
+        Status = data['Status']
+        Comment = data['Comment'] 
+
+
+        new_flag = RedFlagModel(data)
+        resp = self.db.save(new_flag)
+        success_message={
```
Comment (giddygitau) [inline app/api/v1/views.py]: You don't need this success_message dictionary. Just pass `new_flag` as `data`

File: app/api/v1/views.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    def __init__(self):
+        self.db = RedFlagModel()      
+
+    def get(self, Id): 
+        Flag =self.db.viewOne(Id)
+        if not Flag: 
+            return make_response(jsonify({"Flag does not exist":404}), 404)
+        return make_response(jsonify({
+            "Status" : 200,
+            "your entry is": Flag
```
Comment (giddygitau) [inline app/api/v1/views.py]: Change "your entry is" to "data"

File: app/api/v1/views.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        }),200)
+        
+    def delete(self, Id):
+        Flag = self.db.deleteOne(Id)
+        return make_response(jsonify({
+            "Status" : 200,
+            "successfully deleted": self.db
```
Comment (giddygitau) [inline app/api/v1/views.py]: I don't think we should return self.db on delete. Just return a success message or better yet return nothing with the status code 204

File: app/api/v1/views.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    def patch( self, Id ):
+        Flag =self.db.viewOne(Id)
+        if not Flag: 
+            return make_response(jsonify({"Flag does not exist":404}), 404)
+        else:
+            data = Flag.update(request.get_json()) 
+            return make_response(jsonify({
+                "status":200,
+                "Comment edited successfuly":Flag
```
Comment (giddygitau) [inline app/api/v1/views.py]: I think we should be returning `data` instead of `Flag` here. With the property set to data instead of "Comment edited successfully"

**Review by giddygitau — CHANGES_REQUESTED** [review]

File: app/api/v1/views.py (hunk 1 above)
Comment (Muthembwa) [reply app/api/v1/views.py]: I cant figure out the line I have used 1 instead of 4. or should i just transfer the parameters at line 10 to line 14 at post?

File: app/api/v1/views.py (hunk 2 above)
Comment (Muthembwa) [reply app/api/v1/views.py]: Noted.

Comment (Muthembwa) [reply app/api/v1/views.py]: Is it fine the way I pass data to the new flag instance?

File: app/api/v1/views.py (hunk 3 above)
Comment (Muthembwa) [reply app/api/v1/views.py]: Noted

File: app/api/v1/views.py (hunk 4 above)
Comment (Muthembwa) [reply app/api/v1/views.py]: Noted

File: app/api/v1/views.py (hunk 2 above)
Comment (giddygitau) [reply app/api/v1/views.py]: I think the __init__ method of RedFlagModel expects 4 parameters, but you are giving it one. I think this is what is causing the error

File: app/api/v1/views.py (hunk 5 above)
Comment (Muthembwa) [reply app/api/v1/views.py]: Thank you, I will return data Instead, for sure we just need data updated rather than the flag itself.

File: app/api/v1/views.py [hunk 6]
```diff
@@ -31,29 +41,27 @@ def post(self):
...
+        self.db = RedFlagModel()      
 
     def get(self, Id): 
         Flag =self.db.viewOne(Id)
         if not Flag: 
             return make_response(jsonify({"Flag does not exist":404}), 404)
         return make_response(jsonify({
             "Status" : 200,
-            "your entry is": Flag
+            "data": Flag
```
Comment (Muthembwa) [inline app/api/v1/views.py:48]: I have Implemented the suggestions you provided here

File: app/api/v1/views.py [hunk 7]
```diff
@@ -31,29 +41,27 @@ def post(self):
     def get(self):
         resp =self.db.view()
         return make_response(jsonify({
-            "Status" : 200,
-            "All Red Flags":resp
-        }),200)
+            "successfully deleted"
```
Comment (Muthembwa) [inline app/api/v1/views.py]: I thought this suggestion could work in a few other scenarios that I don't necessarily need success messages.

File: app/api/v1/views.py [hunk 8]
```diff
@@ -2,27 +2,37 @@
...
+from .models import RedFlagModel
 
+new_flag = RedFlagModel()
 
-class RedFlags(Resource):
+class RedFlags(Resource, RedFlagModel):
     def __init__(self):
-        self.db = IncidenceModel()
-         
+        self.db = RedFlagModel(CreatedBy, Location, Status, Comment)
```
Comment (Muthembwa) [inline app/api/v1/views.py]: This solves the "missing 4 parameters error" and starts raising "CreatedBy not defined", Can you assist me on actionable steps that can solve this? I will also appreciate any resources that can enhance my understanding specifically on OOP scenarios as the one above.
